src/WebJobs.Extensions.DurableTask/Auth/AzureCredentialFactory.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Microsoft.Extensions.Azure;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage.Auth;
+using AppAuthTokenCredential = Microsoft.WindowsAzure.Storage.Auth.TokenCredential;
+using AzureTokenCredential = Azure.Core.TokenCredential;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Auth
+{
+    internal class AzureCredentialFactory : ITokenCredentialFactory
+    {
+        private const string LoggerName = "Host.Triggers.DurableTask.Auth";
+
+        private readonly string hubName;
+        private readonly AzureComponentFactory componentFactory;
+        private readonly EndToEndTraceHelper traceHelper;
+
+        public AzureCredentialFactory(
+            IOptions<DurableTaskOptions> options,
+            AzureComponentFactory componentFactory,
+            ILoggerFactory loggerFactory)
+        {
+            this.hubName = options?.Value?.HubName ?? throw new ArgumentNullException(nameof(options));
+            this.componentFactory = componentFactory ?? throw new ArgumentNullException(nameof(componentFactory));
+            this.traceHelper = new EndToEndTraceHelper(loggerFactory?.CreateLogger(LoggerName) ?? throw new ArgumentNullException(nameof(loggerFactory)), false);
+        }
+
+        internal event Action<TokenRenewalState> Renewing;
+
+        internal event Action<NewTokenAndFrequency> Renewed;
+
+        internal event Action<int, NewTokenAndFrequency, Exception> RenewalFailed;
+
+        /// <inheritdoc />
+        public AppAuthTokenCredential Create(IConfiguration configuration, CancellationToken cancellationToken = default) =>
+            this.Create(configuration, TimeSpan.FromMinutes(5), TimeSpan.FromSeconds(30), cancellationToken);
+
+        /// <inheritdoc />
+        public AppAuthTokenCredential Create(IConfiguration configuration, TimeSpan tokenRefreshOffset, TimeSpan tokenRefreshRetryDelay, CancellationToken cancellationToken = default)
+        {
+            var context = new TokenRequestContext(new[] { "https://storage.azure.com/.default" }, null);
+
+            AzureTokenCredential tokenCredential = this.componentFactory.CreateTokenCredential(configuration);
+            AccessToken value = tokenCredential.GetToken(context, default);
```

Comment:
We're using the default token because I completely overlooked it! Good catch!

But the method is synchronous because we ultimately fetch the `CloudStorageAccount` typically in synchronous contexts. For example, it is used when creating the scaler as an implementation of `Microsoft.Azure.WebJobs.Host.Scale.IScaleMonitorProvider.GetMonitor`.